quipucords/api/aggregate_report/model.py

```diff
@@ -0,0 +1,291 @@
+            if not fingerprint.system_purpose:
+                aggregated.missing_system_purpose += 1
+            if not fingerprint.redhat_certs:
+                aggregated.missing_pem_files += 1
+        if has_openshift_source:
+            if fingerprint.cpu_count:
+                aggregated.openshift_cores += fingerprint.cpu_count
+        if has_vcenter_source:
+            vmware_cluster = fingerprint.vm_cluster
+            vmware_host = fingerprint.virtual_host_uuid
+            if vmware_cluster and vmware_host:
+                vmware_vms_by_host[vmware_cluster].append(vmware_host)
+
+    aggregated.os_by_name_and_version = {
+        os_name: dict(os_versions)
+        for os_name, os_versions in os_by_name_and_version.items()
+    }
+
+    aggregated.system_creation_date_average = average_date(system_creation_dates)
+
+    if vmware_vms_by_host:
+        # By collecting *all* the hosts in vmware_vms_by_host, we can effectively
+        # deduplicate them here. We collected the hosts and VMs by their names,
+        # not their UUIDs, which means that if two actually different hosts or VMs
+        # have the same name (e.g. "localhost"), then we might under-count them here.
+        # Justin said this is correct and consistent with how other aspects of our
+        # business count usage by simple hostnames.
+        vmware_vm_count = sum(len(set(items)) for items in vmware_vms_by_host.values())
+        vmware_host_count = len(vmware_vms_by_host)
+        aggregated.vmware_vms = vmware_vm_count
+        aggregated.vmware_vm_to_host_ratio = vmware_vm_count / vmware_host_count
+        aggregated.vmware_hosts = vmware_host_count
+
+
+def _aggregate_from_raw_facts(aggregated: AggregateReport, report: Report) -> None:
+    """
+    Populate the AggregateReport object with data from the given Report's RawFacts.
+
+    Do we really want to build reports directly from raw facts?
+    No, but we don't currently have any better options for some data points.
+    TODO Stop building reports based on raw facts when we have better models.
+    """
+    ansible_hosts_in_database = []
+    ansible_hosts_in_jobs = []
+    openshift_operators_by_kind = defaultdict(int)
+
+    # TODO Try to query RawFacts directly instead of using report.sources.
+    # That should be faster than going through the InspectGroupQuerySet. We would still
+    # need to decode the RawFact.value JSON, but we could filter to only get only the
+    # RawFact.name matches we actually need here, and decoding their .value JSON might
+    # be faster than InspectGroupQuerySet decoding *all* of the rows.
+    inspect_groups = report.sources.all()
+    for inspect_group in inspect_groups:
+        for raw_fact in inspect_group["facts"]:
+            source_type = inspect_group.get("source_type")
+            if source_type == DataSources.NETWORK:
+                if not raw_fact.get("host_done", False):
+                    aggregated.missing_host_done += 1
```

Comment:
I kinda regret revealing to @justinkreft what host done means.
While we can use this metric, I'd rather use the more generic `InspectResult.status` and count total failures or unreachable. That should be the proper way to deal with this.

Alternatively, use the counts present on `Task` with type="inspection"

Reply:
We finally discussed this issue during backlog refinement today, and the team with Justin's blessing agreed that we _**SHOULD NOT**_ be using `host_done` in report outputs like this. 🎉  I will change this code to use `InspectResult.status` instead.

Thank you for persistently asking about this. 🙂  The `status` field will be a much more reliable and future-proof option here, and I was only using `host_done` because of the prior insistence that _specifically it_ must be included in the output.